### PR TITLE
Destinations V2 - T&D Streams in parallel

### PR DESCRIPTION
**Pull request by jbfbell**
We were running typing and deduping operations per stream sequentially. This change makes them execute concurrently.

@evantahler @edgao @cynthiaxyin : One thing I have not had time to track down is why snowflake does not show queries in the logs despite seemingly doing the tasks.

In my testing with source faker this was a little bit faster, the snowflake snyc that was taking 45-49 seconds took 30-37 seconds and the bigquery sync that was taking 1m40-1m45 seconds was taking 1m30 seconds

I have not yet set up a test with many streams, which is what exacerbates the sequential T&D issue - will set this up tomorrow.

**UPDATE**

After testing changes locally for BigQuery GCS Staging, I noticed a minimal increase, 25min -> 20 minutes on 30 streams. By removing incremental typing and deduping I saw a significant speed improvement 25min -> 11 min. This convinces me that we should hold off on incremental typing and deduping until we have a better story around it.

While testing snowflake, I encountered several OOM errors, however the syncs do eventually succeed. I saw the same behavior with the non parallel T&D so I'm inclined to think this is an existing bug. 

I created a [platform PR](https://github.com/airbytehq/airbyte-platform-internal/pull/8653) which allows us to configure the number of threads used by typing and deduping, but it will fallback to 8 if nothing is provided.

**SECOND UPDATE**
The tests for Standard Inserts were failing after this change, so I've removed this for standard inserts. SI is already slower than GCS staging so I think this is acceptable

### CODE REVIEW

File: airbyte-integrations/bases/bases-destination-jdbc/src/main/java/io/airbyte/integrations/destination/staging/GeneralStagingFunctions.java [hunk 1]
```diff
@@ -86,6 +91,37 @@ public static void copyIntoTableFromStage(final JdbcDatabase database,
     }
   }
 
+  private static CompletableFuture<Optional<Exception>> dropStageTask(final WriteConfig writeConfig,
+                                                                      final StagingOperations stagingOperations,
+                                                                      final boolean purgeStagingData,
+                                                                      final JdbcDatabase database) {
+    return CompletableFuture.supplyAsync(() -> {
```
Comment (edgao) [inline airbyte-integrations/bases/bases-destination-jdbc/src/main/java/io/airbyte/integrations/destination/staging/GeneralStagingFunctions.java:98]: how many threads does this use? javadoc says it depends on ForkJoinPool.commonPool, which google suggests is limited to `cpu_count - 1` threads, which is probably pretty restrictive on cloud (where we set cpu=3 or something for snowflake?)

I suspect we need to supply our own thread pool. dbt defaults to 4 threads, but we can probably go even higher

**Review by edgao — CHANGES_REQUESTED** [review]
supplyAsync makes me nervous, otherwise I think this looks right. Had a question around correctness wrt to the final dropStage / T+D  interaction, plus a few nits.

File: airbyte-integrations/bases/bases-destination-jdbc/src/main/java/io/airbyte/integrations/destination/staging/GeneralStagingFunctions.java (hunk 1 above)
Comment (evantahler) [reply airbyte-integrations/bases/bases-destination-jdbc/src/main/java/io/airbyte/integrations/destination/staging/GeneralStagingFunctions.java:98]: I also like the idea of us providing the number explicitly.  e.g. in the future, this number should scale with the size of your compute warehouse.  For now, a contestant (4?) is probably a good stating point

**Review by edgao — APPROVED** [review]

**Comment by edgao** [comment]
previous commit timed out on bigquery integration tests. I think we need to call `cleanup` in the standard inserts record consumer as well. will push that commit in a few minutes.

**Comment by jbfbell** [comment]
> previous commit timed out on bigquery integration tests. I think we need to call `cleanup` in the standard inserts record consumer as well. will push that commit in a few minutes.

:doh: good catch, thanks @edgao

File: airbyte-integrations/bases/bases-destination-jdbc/src/main/java/io/airbyte/integrations/destination/staging/GeneralStagingFunctions.java [hunk 2]
```diff
@@ -104,6 +104,7 @@ public static OnCloseFunction onCloseFunction(final JdbcDatabase database,
       // After moving data from staging area to the target table (airybte_raw) clean up the staging
       // area (if user configured)
       log.info("Cleaning up destination started for {} streams", writeConfigs.size());
+      typerDeduper.typeAndDedupe();
```
Comment (edgao) [inline airbyte-integrations/bases/bases-destination-jdbc/src/main/java/io/airbyte/integrations/destination/staging/GeneralStagingFunctions.java:107]: @benmoriceau / @tryangul fyi we're tweaking the T+D interface again 😅 . There's now a `typeAndDedupe()` method, which runs T+D on all streams, along with a new `cleanup` method (under the hood, it shuts down a thread pool, and is mandatory to run at the end of a sync to allow the process to exit)

**Comment by edgao** [comment]
ran a test to dest-bigquery with 11 streams and got this error during table setup:
```
Exceeded rate limits: too many dataset metadata update operations for this dataset.
```
... which I had to go into the bq cli for, because `job.waitFor()` actually never terminated, so the sync just got stuck. Very exciting.

**Comment by edgao** [comment]
snowflake prerelease publish running here https://github.com/airbytehq/airbyte/actions/runs/6090535400

**Pull request merged**